src/web/rustdoc.rs

```diff
@@ -223,20 +223,34 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
         req_path.remove(0);
     }
 
-    version = match match_version(&conn, &name, url_version).and_then(|m| m.assume_exact()) {
-        Some(MatchSemver::Exact((v, _))) => v,
-        Some(MatchSemver::Semver((v, _))) => {
-            // to prevent cloudfront caching the wrong artifacts on URLs with loose semver
-            // versions, redirect the browser to the returned version instead of loading it
-            // immediately
-            let url = ctry!(Url::parse(
-                &format!("{}/{}/{}/{}", base, name, v, req_path.join("/"))[..]
-            ));
-            return Ok(super::redirect(url));
+    let corrected_name;
+    version = match match_version(&conn, &name, url_version) {
+        Some(mv) => {
+            corrected_name = mv.corrected_name;
+
+            match mv.version {
+                MatchSemver::Exact((v, _)) => v,
+                MatchSemver::Semver((v, _)) => {
+                    // to prevent cloudfront caching the wrong artifacts on URLs with loose semver
+                    // versions, redirect the browser to the returned version instead of loading it
+                    // immediately
+                    let url = ctry!(Url::parse(
+                        &format!("{}/{}/{}/{}", base, name, v, req_path.join("/"))[..]
+                    ));
+                    return Ok(super::redirect(url));
+                }
+            }
         }
-        _ => return Err(IronError::new(Nope::ResourceNotFound, status::NotFound)),
+        None => return Err(IronError::new(Nope::ResourceNotFound, status::NotFound)),
     };
 
+    if let Some(name) = corrected_name {
+        let url = ctry!(Url::parse(
+            &format!("{}/{}/{}/{}", base, name, version, req_path.join("/"))[..]
+        ));
+        return Ok(super::redirect(url));
+    }
```

Comment:
Hmm, it might be worth making a helper function for this, I think we use it in a couple other places as well.